src/components/spreadsheet/custom-columns/use-custom-column.ts

```diff
@@ -0,0 +1,151 @@
+        const limitedEvaluate = instance.evaluate.bind(instance);
+        // Disable potentially dangerous functions
+        instance.import(
+            {
+                import: function () {
+                    throw new Error('Function import is disabled');
+                },
+                createUnit: function () {
+                    throw new Error('Function createUnit is disabled');
+                },
+                evaluate: function () {
+                    throw new Error('Function evaluate is disabled');
+                },
+                parse: function () {
+                    throw new Error('Function parse is disabled');
+                },
+                simplify: function () {
+                    throw new Error('Function simplify is disabled');
+                },
+                derivative: function () {
+                    throw new Error('Function derivative is disabled');
+                },
+                equal: function (a: any, b: any) {
+                    // == instead of === to be able to compare strings to numbers
+                    return a === b;
+                },
+            },
+            { override: true }
+        );
+        return { limitedEvaluate };
+    }, []);
+
+    const createDependencyGraph = (customColumnDefinitions: ColumnWithFormula[]): Map<string, Set<string>> => {
+        const graph = new Map<string, Set<string>>();
+
+        // Initialize the graph with all columns
+        customColumnDefinitions.forEach((col) => {
+            graph.set(col.name, new Set());
+        });
+
+        // Build the graph by adding dependencies between columns
+        customColumnDefinitions.forEach((col) => {
+            customColumnDefinitions.forEach((depCol) => {
+                if (col.formula.includes(depCol.name) && col.name !== depCol.name) {
+                    graph.get(col.name)!.add(depCol.name);
+                }
+            });
+        });
+
+        return graph;
+    };
+
+    const topologicalSort = (graph: Map<string, Set<string>>): string[] => {
+        const sorted: string[] = [];
+        const visited = new Set<string>();
+        const visiting = new Set<string>(); // Temporary set for detecting cycles
+
+        const dfs = (node: string) => {
+            if (visiting.has(node)) {
+                throw new Error(`Circular dependency detected at column: ${node}`);
+            }
+            if (!visited.has(node)) {
+                visiting.add(node); // Mark the node as currently being visited
+                const dependencies = graph.get(node) || new Set();
+                dependencies.forEach(dfs);
+                visiting.delete(node);
+                visited.add(node);
+                sorted.push(node);
+            }
+        };
+
+        // Start DFS from each node
+        for (const node of graph.keys()) {
+            if (!visited.has(node)) {
+                dfs(node);
+            }
+        }
+
+        return sorted;
+    };
```

Comment:
You could extract that in a utility file that will not enter in react processes of rendering

Reply:
Done